task2/rest-api-basics/service/src/main/java/com/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,230 @@
+package com.epam.esm.service.impl;
+
+import com.epam.esm.constant.ServiceError;
+import com.epam.esm.dto.GiftCertificateDto;
+import com.epam.esm.converter.Converter;
+import com.epam.esm.entity.GiftCertificate;
+import com.epam.esm.entity.Tag;
+import com.epam.esm.repository.GiftCertificateRepository;
+import com.epam.esm.repository.TagRepository;
+import com.epam.esm.repository.exception.RepositoryException;
+import com.epam.esm.service.GiftCertificateService;
+import com.epam.esm.service.exception.certificate.GiftCertificateAlreadyExistException;
+import com.epam.esm.service.exception.certificate.GiftCertificateNotFoundException;
+import com.epam.esm.service.exception.ServiceException;
+import com.epam.esm.service.exception.certificate.UnableDeleteGiftCertificateException;
+import com.epam.esm.service.exception.certificate.UnableUpdateGiftCertificate;
+import com.epam.esm.util.ParamsUtil;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.time.LocalDateTime;
+import java.util.*;
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
```

Comment:
Please use @Transactiona(readOnly = true) annotation for read operations too.
![image](https://user-images.githubusercontent.com/32880986/104951425-84728000-59d3-11eb-9a99-709ae0991f0c.png)
https://stackoverflow.com/questions/44984781/what-are-advantages-of-using-transactionalreadonly-true

Reply:
thanks